### PR TITLE
Multiclass interface

### PR DESCRIPTION
**Pull request by colbybanbury**
Initial version of the multiclass interface.
**Do not Merge** yet

TODO:

- README updates
- Clean up of code (move file access to functions to simplify flow)
- Dockerize selection
- Tune models

### CODE REVIEW

File: eval.py [hunk 1]
```diff
@@ -2,47 +2,108 @@
...
-    eval_nontarget = eval_data["nontarget_mswc_vectors"]
-    eval_x = np.vstack([eval_target, eval_nontarget])
-    eval_y = np.concatenate(
-        [np.ones(eval_target.shape[0]), np.zeros(eval_nontarget.shape[0])]
-    )
+def create_train_sets(selected_data, allowed_training_embeddings, training_classes, train_dataset):
+    train_x = [] #embeddings
+    train_y = [] #numerical index of training_classes 
+    for selected_id in selected_data["nontargets"]:
+        if not selected_id in allowed_training_embeddings['nontargets']:
```
Comment (mmaz) [inline eval.py:16]:
```suggestion
    train_y = [] #numerical index of training_classes 
    allowed_nontargets = set(allowed_training_embeddings['nontargets'])
    for selected_id in selected_data["nontargets"]:
        if not selected_id in allowed_nontargets:
```

alternatively we could setify it in the yaml loader

File: eval.py [hunk 2]
```diff
@@ -2,47 +2,108 @@
...
+            print("invaid embedding ID: ", selected_id)
+            raise ValueError('Selected Embedding ID not present in allowed training data')
 
-    sample_data = deserialize.deserialize_from_pb(input_samples)
-    target_vectors = sample_data["target_mswc_vectors"]
-    target_ids = sample_data["target_ids"]
-    nontarget_vectors = sample_data["nontarget_mswc_vectors"]
-    nontarget_ids = sample_data["nontarget_ids"]
+        label = Path(selected_id).parts[label_idx]
```
Comment (mmaz) [inline eval.py:20]: i'm confused - if idx is 0, doesn't it return `en`? (in `en/clips/and/common_voice_12345.opus`)

File: eval.py [hunk 3]
```diff
@@ -2,47 +2,108 @@
...
+
+    with open(config_file, "r") as fh:
+        config = yaml.safe_load(fh)
+    train_set_size_limit = config["train_set_size_limit"]
+    #TODO add more config parameters
+    
+    with open(allowed_training_set, "r") as fh:
+        allowed_training_embeddings = yaml.safe_load(fh) #dict {"targets": {"dog":[list]}, "nontargets": [list]}
+    training_classes = list(allowed_training_embeddings["targets"].keys())#specifies the label indicies for training 
+    training_classes.append("nontarget")
```
Comment (mmaz) [inline eval.py:79]: this is just a stylistic choice but what do you think about prepending the list so that `nontarget` is always id `0`?

File: eval.py [hunk 4]
```diff
@@ -2,47 +2,108 @@
...
+            config_file="dataperf_speech_config.yaml"):
+
+    with open(config_file, "r") as fh:
+        config = yaml.safe_load(fh)
+    train_set_size_limit = config["train_set_size_limit"]
+    #TODO add more config parameters
+    
+    with open(allowed_training_set, "r") as fh:
+        allowed_training_embeddings = yaml.safe_load(fh) #dict {"targets": {"dog":[list]}, "nontargets": [list]}
+    training_classes = list(allowed_training_embeddings["targets"].keys())#specifies the label indicies for training
```
Comment (mmaz) [inline eval.py:78]:
```suggestion
    training_classes = list(allowed_training_embeddings["targets"].keys())#specifies the label indicies for training 
    assert "nontarget" not in training_classes, "nontarget is a reserved label"
```

File: eval.py [hunk 5]
```diff
@@ -2,47 +2,108 @@
...
+        parquet_file = pd.read_parquet(train_dataset / (label + ".parquet"))
+        idx = parquet_file["clip_id"].tolist().index(selected_id)
+        train_x.append(parquet_file["mswc_embedding_vector"][idx])
+        train_y.append(training_classes.index("nontarget"))
+
+    for label in selected_data["targets"].keys():
+        if not label in training_classes:
+            raise ValueError('Provided label doesnt match classes in eval set')
+        for selected_id in selected_data["targets"][label]:
+            if not selected_id in allowed_training_embeddings['targets'][label]:
```
Comment (mmaz) [inline eval.py:33]: we can setify this too either above the for loop or in the loader

File: eval.py [hunk 6]
```diff
@@ -2,47 +2,108 @@
...
+
+    train_x, train_y = create_train_sets(selected_data, allowed_training_embeddings, training_classes, train_dataset)
+        
+    if len(train_y) > train_set_size_limit:
+        print("Selected training set size: ", len(train_y))
+        print("Training set size limit: ", train_set_size_limit)
+        raise ValueError('Selected training set is too large')
+
+
+    seed = 0
```
Comment (mmaz) [inline eval.py:95]: we should load this in from the yaml config file

File: requirements.txt [hunk 7]
```diff
@@ -1,4 +1,5 @@
-# yacs>=0.1.8
 pyyaml>=5.4.1
 fire
-protobuf
\ No newline at end of file
+pandas
+pyarrow
+fastparquet
```
Comment (mmaz) [inline requirements.txt:5]: i think we only want one of the two? (I'd go with pyarrow)

File: selection/selection.py [hunk 8]
```diff
@@ -40,20 +35,23 @@ def select(self):
...
+
+        per_class_size = self.train_set_size // 6 #5 targets + nontarget
+
+        selected_targets = {}
+        for target, sample_list in self.target_vectors['targets'].items():
+            selected_samples = sample(sample_list, per_class_size)
+            selected_targets[target] = [sample['ID'] for sample in selected_samples]
 
-        rng = np.random.RandomState(0)
+        selected_nontarget_samples = sample(self.target_vectors['non_targets'], per_class_size)
```
Comment (mmaz) [inline selection/selection.py:48]: I think we will want to add in the random seed from the config and pass that in as a constructor argument, so we might as well use numpy here (especially to encourage future participants to be as deterministic as possible for repeatability)

File: selection/selection.py [hunk 9]
```diff
@@ -1,36 +1,31 @@
...
+        train_set_size,
+        audio_flag=False
     ) -> None:
-        """
-        Arguments:
-            target_vectors:
-            target_ids:
-            nontarget_vectors:
-            nontarget_ids:
-        """
```
Comment (mmaz) [inline selection/selection.py:29]: i think we should keep this docstring placeholder, since we need to fill everything out with docstrings before sharing the alpha (fire will also use our main function docstrings for outputting commandline help)

File: Dockerfile [hunk 10]
```diff
@@ -1,9 +1,10 @@
...
-COPY requirements.txt /app/requirements.txt
-RUN pip install -r /app/requirements.txt
+COPY requirements.txt ./requirements.txt
+RUN pip install --no-cache-dir -r ./requirements.txt
 
-COPY . /app/
-WORKDIR /app/
+COPY selection/ ./selection
+COPY dataperf_speech_config.yaml ./dataperf_speech_config.yaml
+COPY allowed_training_set.yaml ./allowed_training_set.yaml
```
[the remark this answers is not in the document]
Comment (colbybanbury) [reply Dockerfile:8]: Good point!

Easy fix for the .yaml files but I'm not sure how to copy in selection.py while mounting  /selection on run.

Any advice?

File: eval.py (hunk 2 above)
Comment (colbybanbury) [reply eval.py:21]: the IDs in the parquet files are actually of the form `left/common_voice_en_2145.wav`

I made it a variable cause I thought that might change

File: Dockerfile (hunk 10 above)
[the remark this answers is not in the document]
Comment (mmaz) [reply Dockerfile:8]: the idea in my mind is to copy the whole directory in (hence why we should explicitly discourage downloading the embeddings inside the repo dir, otherwise the image will be 30gb) so selection.py will already be present in the built container... does that not work?

File: eval.py (hunk 3 above)
Comment (colbybanbury) [reply eval.py:80]: Yeah I agree that feels right

File: eval.py (hunk 2 above)
Comment (colbybanbury) [reply eval.py:21]: oh right! Good catch lol

**Comment by mmaz** [comment]
thanks!
